drivers/ethernet/eth_numaker.c

```diff
@@ -0,0 +1,807 @@
+#ifdef CONFIG_SOC_M467
+#include <m460_eth.h>
+#endif
+
+LOG_MODULE_REGISTER(eth_numaker, CONFIG_ETHERNET_LOG_LEVEL);
+
+/* Device EMAC Interface port */
+#define NUMAKER_GMAC_INTF  0
+/* 2KB Data Flash at 0xFF800 */
+#define NUMAKER_DATA_FLASH (0xFF800U)
+#define NUMAKER_MASK_32    (0xFFFFFFFFU)
+extern synopGMACdevice GMACdev[GMAC_CNT];
+extern struct sk_buff tx_buf[GMAC_CNT][TRANSMIT_DESC_SIZE];
+extern struct sk_buff rx_buf[GMAC_CNT][RECEIVE_DESC_SIZE];
+
+static uint32_t eth_phy_addr;
+
+/* Device config */
+struct eth_numaker_config {
+	/* GMAC Base address */
+	uint32_t gmac_base;
+	const struct reset_dt_spec reset;
+	uint32_t phy_addr;
+	/* clock configuration */
+	uint32_t clk_modidx;
+	uint32_t clk_src;
+	uint32_t clk_div;
+	const struct device *clk_dev;
+	const struct pinctrl_dev_config *pincfg;
+};
+
+/* Driver context/data */
+struct eth_numaker_data {
+	synopGMACdevice *gmacdev;
+	struct net_if *iface;
+	uint8_t mac_addr[NU_HWADDR_SIZE];
+	struct k_mutex tx_frame_buf_mutex;
+	struct k_spinlock rx_frame_buf_lock;
+};
+
+/* Delay execution for given amount of ticks for SDK-HAL */
+void plat_delay(uint32_t delay)
+{
+	uint32_t us_cnt = k_ticks_to_us_floor32((uint64_t)delay);
+
+	k_busy_wait(us_cnt);
+}
+
+static void mdio_write(synopGMACdevice *gmacdev, uint32_t addr, uint32_t reg, int data)
+{
+	synopGMAC_write_phy_reg((u32 *)gmacdev->MacBase, addr, reg, data);
+}
+
+static int mdio_read(synopGMACdevice *gmacdev, uint32_t addr, uint32_t reg)
+{
+	uint16_t data;
+
+	synopGMAC_read_phy_reg((u32 *)gmacdev->MacBase, addr, reg, &data);
+	return data;
+}
+
+static int numaker_eth_link_ok(synopGMACdevice *gmacdev)
+{
+	/* first, a dummy read to latch */
+	mdio_read(gmacdev, eth_phy_addr, MII_BMSR);
+	if (mdio_read(gmacdev, eth_phy_addr, MII_BMSR) & BMSR_LSTATUS) {
+		return 1;
+	}
+	return 0;
+}
+
+static int reset_phy(synopGMACdevice *gmacdev)
+{
+	uint16_t reg;
+	uint32_t delay_cnt;
+	int status = 0;
+	bool ret;
+
+	mdio_write(gmacdev, eth_phy_addr, MII_BMCR, BMCR_RESET);
+
+	delay_cnt = 200000U;
+	ret = WAIT_FOR(!(mdio_read(gmacdev, eth_phy_addr, MII_BMCR) & BMCR_RESET),
+					delay_cnt, k_msleep(1));
+
+	if (ret == false) {
+		LOG_DBG("Reset phy failed");
+		return -ENODATA;
+	}
+
+	delay_cnt = 3000000U;
+	ret = WAIT_FOR(numaker_eth_link_ok(gmacdev), delay_cnt, k_msleep(1));
+
+	if (ret) {
+		gmacdev->LinkState = LINKUP;
+		LOG_DBG("Link Up");
+	}
+
+	if (ret == false) {
+		gmacdev->LinkState = LINKDOWN;
+		LOG_DBG("Link Down");
+		return -ENODATA;
+	}
+
+	mdio_write(gmacdev, eth_phy_addr, MII_ADVERTISE,
+		   ADVERTISE_CSMA | ADVERTISE_10HALF | ADVERTISE_10FULL | ADVERTISE_100HALF |
+			   ADVERTISE_100FULL);
+
+	reg = mdio_read(gmacdev, eth_phy_addr, MII_BMCR);
+	mdio_write(gmacdev, eth_phy_addr, MII_BMCR, reg | BMCR_ANRESTART);
+
+	delay_cnt = 3000000U;
+	ret = WAIT_FOR((mdio_read(gmacdev, eth_phy_addr, MII_BMSR) &
+					(BMSR_ANEGCOMPLETE | BMSR_LSTATUS)) ==
+					(BMSR_ANEGCOMPLETE | BMSR_LSTATUS),
+					delay_cnt, k_msleep(1));
+
```

Comment:
nit: the placing of empty lines is a bit awkward around this whole file, normally the error conditions are just after the function that set the error and there'a blank line after, I think it would be easier to follow if you also do that, for example in this case you could drop this blank and add one  after the `if () {}` block below.